Specs/GCM.lean

```diff
@@ -0,0 +1,198 @@
+/-
+Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author(s): Yan Peng
+-/
+import Arm.BitVec
+import Arm.Insts.Common
+
+-- References: https://nvlpubs.nist.gov/nistpubs/legacy/sp/nistspecialpublication800-38d.pdf
+
+namespace GCM
+
+open BitVec
+
+def R : (BitVec 128) := 0b11100001#8 ++ 0b0#120
+
+/-- A Cipher type is a function that takes an input of size n and
+    a key of size m and returns an encrypted block of size n -/
+abbrev Cipher {n : Nat} {m : Nat} :=  BitVec n → BitVec m → BitVec n
+
+/-- The s-bit incrementing function -/
+def inc_s (s : Nat) (X : BitVec l) (H₀ : 0 < s) (H₁ : s < l) : BitVec l :=
+  let msb_hi := l - 1
+  let msb_lo := s
+  let lsb_hi := s - 1
+  let lsb_lo := 0
+  have h₁ : lsb_hi - lsb_lo + 1 = s := by omega
+  let upper := extractLsb msb_hi msb_lo X
+  let lower := h₁ ▸ (extractLsb lsb_hi lsb_lo X) + 0b1#s
+  have h₂ : msb_hi - msb_lo + 1 + s = l := by omega
+  h₂ ▸ (upper ++ lower)
+
+def mul_aux (i : Nat) (X : BitVec 128) (Z : BitVec 128) (V : BitVec 128)
+  : BitVec 128 :=
+  if h : 128 ≤ i then
+    Z
+  else
+    let Xi := getMsb X i
```

Comment:
Please correct me if I'm wrong, but are we using a different endianness than in the spec here? The spec uses `Xi = getLsb X i`, or not?

Reply:
No, I got this wrong, sorry. The indexing is a bit confusing in the spec since bitstrings are interpreted as little-endian polynomials, which makes the LSB of a bitstring the highest monomial in the represented polynomial.